docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,424 @@
+Some examples of an `on-demand/out-of-band` tasks:
+
+* Recover from a permanent quorum loss of etcd cluster.
+* Trigger an on-demand full/delta snapshot.
+* Trigger an on-demand snapshot compaction.
+* Trigger an on-demand maintenance of etcd cluster.
+* Copy the backups from one object store to another object store.
+
+## Goals
+
+* Establish a unified interface for operator tasks by defining a single dedicated custom resource for `out-of-band` tasks.
+* Define a contract (in terms of prerequisites) which needs to be adhered to by any task implementation.
+* Facilitate the easy addition of new `out-of-band` task(s) through this custom resource.
+* Provide CLI capabilities to operators, making it easy to invoke supported `out-of-band` tasks.
+
+## Non-Goals
+
+* In the current scope, capability to abort/suspend an `out-of-band` task is not going to be provided. This could be considered as an enhancement based on pull.
+* Ordering (by establishing dependency) of `out-of-band` tasks submitted for the same etcd cluster has not been considered in the first increment. In a future version based on how operator tasks are used, we will enhance this proposal and the implementation.
+
+## Proposal
+
+Authors propose creation of a new single dedicated custom resource to represent an `out-of-band` task. Etcd-druid will be enhanced to process the task requests and update its status which can then be tracked/observed.
+
+### Custom Resource API
+
+`EtcdOperatorTask` is the new custom resource that will be introduced. This API will be in `v1alpha1` version and will be subject to change. We will be respecting [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
+
+```go
+// EtcdOperatorTask represents an out-of-band operator task resource.
+type EtcdOperatorTask struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // Spec is the specification of the EtcdOperatorTask resource.
+  Spec EtcdOperatorTaskSpec `json:"spec"`
+  // Status is most recently observed status of the EtcdOperatorTask resource.
+  Status EtcdOperatorTaskStatus `json:"status,omitempty"`
+}
+```
+
+#### Spec
+
+The authors propose that the following fields should be specified in the spec (desired state) of the `EtcdOperatorTask` custom resource.
+
+* To capture the type of `out-of-band` operator task to be performed, `.spec.type` field should be defined. It can have values from all supported `out-of-band` tasks eg. "OnDemandSnaphotTask", "QuorumLossRecoveryTask" etc.
+* To capture the configuration specific to each task, a `.spec.config` field should be defined of type `string` as each task can have different input configuration.
```

Comment:
```suggestion
* To capture the configuration specific to each task, a `.spec.config` field should be defined. It is expected that the configuration is supplied as JSON string.
```

Reply:
I thought we discussed to remove describing only a subset of fields, right?

Reply:
but I think you already mentioned that in `spec` , we have mentioned only +required fields, which is fine https://github.com/gardener/etcd-druid/pull/757#discussion_r1589007631